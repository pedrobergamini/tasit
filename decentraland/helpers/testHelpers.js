import AssetTypes from "@constants/AssetTypes";
const { ESTATE, PARCEL } = AssetTypes;

export const parcel = {
  type: PARCEL,
  id: "0123456789",
  name: "Sample Parcel",
  img: "https://api.decentraland.org/v1/parcels/0/0/map.png",
};

export const estate = {
  type: ESTATE,
  id: "123",
  name: "Sample Estate",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateForSale = {
  id: "1",
  priceMana: 0,
  priceUSD: 0,
  asset: estate,
};

export const parcelForSale = {
  id: "2",
  priceMana: 0,
  priceUSD: 0,
  asset: parcel,
};

export const anAction = {
  getId: async () =>
    "0x1234567890123456789012345678901234567890123456789012345678901234",
};

export default {
  estateForSale,
  parcelForSale,
  parcel,
  estate,
  anAction,
};
