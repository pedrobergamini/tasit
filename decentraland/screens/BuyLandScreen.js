import React from "react";
import { connect } from "react-redux";
import {
  removeLandForSale,
  prependLandForSaleToList,
  removeMyAssetFromList,
  addToMyAssetsList,
} from "../redux/actions";
import BuyLand from "@presentational/BuyLand";
import PropTypes from "prop-types";
import { showError, showInfo, getContracts } from "@helpers";
import { storeMyAssets } from "@helpers/storage";
import AssetTypes from "@constants/AssetTypes";
const { ESTATE, PARCEL } = AssetTypes;

// TODO: Go deep on gas handling.
// Without that, VM returns a revert error instead of out of gas error.
// See: https://github.com/tasitlabs/TasitSDK/issues/173
const gasParams = {
  gasLimit: 7e6,
  gasPrice: 1e9,
};

export class BuyLandScreen extends React.Component {
  _onBuy = landForSale => {
    try {
      const { accountInfo } = this.props;
      const { account } = accountInfo;
      if (!account) this._setupAccount();
      // The _buy function assumes that the account is created, funded and allowed
      // The user will get back to this else case after going through an account setup flow
      else this._buy(landForSale);
    } catch (err) {
      showError(err);
    }
  };

  _setupAccount = () => {
    const { navigation } = this.props;
    navigation.navigate("OnboardingHomeScreen");
  };

  _buy = landForSale => {
    const { props, _executeOrder } = this;
    const {
      navigation,
      accountInfo,
      myAssets,
      removeLandForSale,
      prependLandForSaleToList,
      removeMyAssetFromList,
      addToMyAssetsList,
    } = props;
    const { account } = accountInfo;
    const { asset } = landForSale;
    const { type } = asset;

    if (type !== ESTATE && type !== PARCEL) showError(`Unknown asset.`);

    const typeDescription = type == ESTATE ? "Estate" : "Parcel";

    const onSuccess = async action => {
      showInfo(`${typeDescription} bought successfully.`);

      const actionId = await action.getId();
      const boughtAsset = { ...asset, actionId };
      storeMyAssets([boughtAsset, ...myAssets]);
    };

    const onError = (assetForSale, message) => {
      const { asset } = assetForSale;
      showError(message);
      removeMyAssetFromList(asset);
      prependLandForSaleToList(assetForSale);
    };

    showInfo(`Buying the ${typeDescription.toLowerCase()}...`);

    // Optimistic UI update
    removeLandForSale(landForSale);
    addToMyAssetsList(asset);

    _executeOrder(landForSale, account, onSuccess, onError);
    navigation.navigate("ListLandForSaleScreen");
  };

  _executeOrder = async (
    sellOrder,
    account,
    afterSuccessfulExecution,
    onError
  ) => {
    try {
      const { priceManaInWei: priceInWei, asset } = sellOrder;
      const { id: assetId, type } = asset;
      const contracts = getContracts();
      const { marketplaceContract, estateContract, landContract } = contracts;

      let nftAddress;
      let fingerprint;
      if (type == ESTATE) {
        nftAddress = estateContract.getAddress();
        fingerprint = await estateContract.getFingerprint(assetId);
      } else if (type == PARCEL) {
        nftAddress = landContract.getAddress();
        // LANDRegistry contract doesn't implement getFingerprint function
        fingerprint = "0x";
      }

      marketplaceContract.setWallet(account);
      const action = marketplaceContract.safeExecuteOrder(
        nftAddress,
        `${assetId}`,
        `${priceInWei}`,
        `${fingerprint}`,
        gasParams
      );

      // TODO: This function should be called inside of the eventListener
      // that catches the safeExecuteOrder successful event.
      await action.waitForNonceToUpdate();

      afterSuccessfulExecution(action);
    } catch (error) {
      // Note: The current version isn't supporting `failing` events
      // See more:
      // https://github.com/tasitlabs/tasit/issues/151
      // https://github.com/tasitlabs/tasit/issues/233
      onError(sellOrder, error.message);
    }
  };

  render() {
    const { selectedLandToBuy: landForSale, accountInfo } = this.props;
    const { creationStatus, creationActions } = accountInfo;

    return (
      <BuyLand
        landForSale={landForSale}
        onBuy={() => this._onBuy(landForSale)}
        accountCreationStatus={creationStatus}
        accountCreationActions={creationActions}
      />
    );
  }
}

BuyLandScreen.propTypes = {
  accountInfo: PropTypes.object,
  selectedLandToBuy: PropTypes.object.isRequired,
  myAssets: PropTypes.array.isRequired,
  removeLandForSale: PropTypes.func.isRequired,
  prependLandForSaleToList: PropTypes.func.isRequired,
  removeMyAssetFromList: PropTypes.func.isRequired,
  addToMyAssetsList: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  const { accountInfo, selectedLandToBuy, myAssets } = state;
  const { list: myAssetsList } = myAssets;
  return { accountInfo, selectedLandToBuy, myAssets: myAssetsList };
};

const mapDispatchToProps = {
  removeLandForSale,
  prependLandForSaleToList,
  addToMyAssetsList,
  removeMyAssetFromList,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BuyLandScreen);
