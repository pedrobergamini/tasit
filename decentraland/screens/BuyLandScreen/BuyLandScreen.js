import React from "react";
import { connect } from "react-redux";
import { StackActions } from "react-navigation";
import {
  removeLandForSale,
  prependLandForSaleToList,
  removeFromMyAssetsList,
  prependToMyAssetsList,
  addUserAction,
  updateUserActionStatus,
} from "@redux/actions";
import BuyLand from "@presentational/BuyLand";
import PropTypes from "prop-types";
import { showError, showInfo, getContracts } from "@helpers";
import { ESTATE, PARCEL } from "@constants/AssetTypes";
import { PENDING, SUCCESSFUL } from "@constants/UserActionStatus";

// TODO: Go deep on gas handling.
// Without that, VM returns a revert error instead of out of gas error.
// See: https://github.com/tasitlabs/TasitSDK/issues/173
const gasParams = {
  gasLimit: 7e6,
  gasPrice: 1e9,
};

export class BuyLandScreen extends React.Component {
  _onBuy = landForSale => {
    try {
      const { accountInfo } = this.props;
      const { account } = accountInfo;
      if (!account) this._setupAccount();
      // The _buy function assumes that the account is created, funded and allowed
      // The user will get back to this else case after going through an account setup flow
      else this._buy(landForSale);
    } catch (err) {
      showError(err);
    }
  };

  _setupAccount = () => {
    const { navigation } = this.props;
    navigation.navigate("OnboardingHomeScreen");
  };

  _buy = async landForSale => {
    const { props, _executeOrder } = this;
    const {
      navigation,
      accountInfo,
      removeLandForSale,
      prependLandForSaleToList,
      removeFromMyAssetsList,
      prependToMyAssetsList,
      addUserAction,
      updateUserActionStatus,
    } = props;
    const { account } = accountInfo;
    const { asset } = landForSale;
    const { id: assetId, type } = asset;

    if (type !== ESTATE && type !== PARCEL) showError(`Unknown asset.`);

    const typeDescription = type == ESTATE ? "Estate" : "Parcel";

    const onSuccess = async () => {
      // TODO: This function should be called inside of the eventListener
      // that catches the safeExecuteOrder successful event.
      await action.waitForNonceToUpdate();
      const actionId = await action.getId();

      updateUserActionStatus({ actionId, status: SUCCESSFUL });

      showInfo(`${typeDescription} bought successfully.`);
    };

    const onError = (assetForSale, message) => {
      const { asset } = assetForSale;
      showError(message);
      removeFromMyAssetsList(asset);
      prependLandForSaleToList(assetForSale);
    };

    showInfo(`Buying the ${typeDescription.toLowerCase()}...`);

    const action = await _executeOrder(landForSale, account, onError);

    // Optimistic UI update
    removeLandForSale(landForSale);
    prependToMyAssetsList(asset);

    // Back to top of current Stack before navigate
    navigation.dispatch(StackActions.popToTop());
    navigation.navigate("MyAssetsScreen");

    const actionId = await action.getId();
    const userAction = { [actionId]: { status: PENDING, assetId } };

    addUserAction(userAction);

    onSuccess();
  };

  _executeOrder = async (sellOrder, account, onError) => {
    try {
      const { priceManaInWei: priceInWei, asset } = sellOrder;
      const { id: assetId, type } = asset;
      const contracts = getContracts();
      const { marketplaceContract, estateContract, landContract } = contracts;

      const nftAddress =
        type === ESTATE
          ? estateContract.getAddress()
          : landContract.getAddress();

      // LANDRegistry contract doesn't implement getFingerprint function
      const fingerprint =
        type === ESTATE ? await estateContract.getFingerprint(assetId) : "0x";

      marketplaceContract.setWallet(account);
      const action = marketplaceContract.safeExecuteOrder(
        nftAddress,
        `${assetId}`,
        `${priceInWei}`,
        `${fingerprint}`,
        gasParams
      );

      return action;
    } catch (error) {
      // Note: The current version isn't supporting `failing` events
      // See more:
      // https://github.com/tasitlabs/tasit/issues/151
      // https://github.com/tasitlabs/tasit/issues/233
      onError(sellOrder, error.message);
    }
  };

  render() {
    const { landToBuy: landForSale, accountInfo } = this.props;
    const { creationStatus, creationActions } = accountInfo;

    return (
      <BuyLand
        landForSale={landForSale}
        onBuy={() => this._onBuy(landForSale)}
        accountCreationStatus={creationStatus}
        accountCreationActions={creationActions}
      />
    );
  }
}

BuyLandScreen.propTypes = {
  accountInfo: PropTypes.object,
  landToBuy: PropTypes.object.isRequired,
  myAssets: PropTypes.array.isRequired,
  removeLandForSale: PropTypes.func.isRequired,
  prependLandForSaleToList: PropTypes.func.isRequired,
  removeFromMyAssetsList: PropTypes.func.isRequired,
  prependToMyAssetsList: PropTypes.func.isRequired,
  addUserAction: PropTypes.func.isRequired,
  updateUserActionStatus: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  const { accountInfo, landToBuy, myAssets } = state;
  const { list: myAssetsList } = myAssets;
  return { accountInfo, landToBuy, myAssets: myAssetsList };
};

const mapDispatchToProps = {
  removeLandForSale,
  prependLandForSaleToList,
  prependToMyAssetsList,
  removeFromMyAssetsList,
  addUserAction,
  updateUserActionStatus,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BuyLandScreen);
