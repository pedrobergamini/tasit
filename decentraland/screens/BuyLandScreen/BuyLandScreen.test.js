import React from "react";
import { shallow } from "enzyme";
import { BuyLandScreen } from "./BuyLandScreen";
import { estateForSale } from "@helpers/testHelpers";
import AccountCreationStatus from "@constants/AccountCreationStatus";
const { READY_TO_USE } = AccountCreationStatus;

describe("BuyLandScreen", () => {
  describe("renders the component", () => {
    it("", async () => {
      const accountInfo = {
        creationStatus: READY_TO_USE,
        creationActions: {},
      };
      const navigation = () => {};
      const removeLandForSale = () => {};
      const prependToMyAssetsList = () => {};
      const removeFromMyAssetsList = () => {};
      const prependLandForSaleToList = () => {};
      const addUserAction = () => {};
      const updateUserActionStatus = () => {};
      const myAssets = [];

      expect(
        shallow(
          <BuyLandScreen
            navigation={navigation}
            accountInfo={accountInfo}
            myAssets={myAssets}
            landToBuy={estateForSale}
            removeLandForSale={removeLandForSale}
            prependToMyAssetsList={prependToMyAssetsList}
            removeFromMyAssetsList={removeFromMyAssetsList}
            prependLandForSaleToList={prependLandForSaleToList}
            addUserAction={addUserAction}
            updateUserActionStatus={updateUserActionStatus}
          />
        )
      ).toMatchSnapshot();
    });
  });
});
