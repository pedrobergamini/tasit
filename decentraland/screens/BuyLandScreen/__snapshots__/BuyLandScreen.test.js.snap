// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyLandScreen renders the component  1`] = `
<BuyLand
  accountCreationActions={Object {}}
  accountCreationStatus="READY_TO_USE"
  landForSale={
    Object {
      "asset": Object {
        "id": "123",
        "img": "https://api.decentraland.org/v1/estates/1/map.png",
        "name": "Sample Estate",
        "type": "ESTATE",
      },
      "id": "1",
      "priceMana": "0",
      "priceUSD": 0,
    }
  }
  onBuy={[Function]}
/>
`;
